core/src/local/mod.rs

```diff
@@ -1,12 +1,51 @@
+use std::borrow::Cow;
 use std::ops::Deref;
 use std::rc::Rc;
 
+mod memory_state;
+mod persistent_state;
 mod single_state;
-mod state;
 mod keyed_state;
 
-pub use data::{DataType, SizeOf};
-pub use self::state::State;
+pub use data::{DataType, Records, SizeOf};
+pub use self::persistent_state::PersistentState;
+pub use self::memory_state::MemoryState;
+
+pub trait State: SizeOf + Send {
+    /// Add an index keyed by the given columns and replayed to by the given partial tags.
+    fn add_key(&mut self, columns: &[usize], partial: Option<Vec<Tag>>);
+
+    /// Returns whether this state is currently keyed on anything. If not, then it cannot store any
+    /// infromation and is thus "not useful".
+    fn is_useful(&self) -> bool;
+
+    fn is_partial(&self) -> bool;
+
+    fn process_records(&mut self, records: &mut Records, partial_tag: Option<Tag>);
```

Comment:
I think it'd be useful to include a doc here about why we give `&mut Records`. The reason is specifically that if the state is partial and a given record misses in all indices in a state, then it should be *removed* from `records`. This is quite important, and I have completely forgotten!